File: CS/AgendViewComponent/AgendaViewMenuBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraScheduler;
using DevExpress.XtraGrid.Menu;
using System.Windows.Forms;
using System.Drawing;
using System.Reflection;

namespace AgendViewComponent {
    public static class AgendaViewMenuBuilder {
        public static SchedulerControl Scheduler { get; set; }
        public static Appointment CurrentAppointment { get; set; }

        public static AgendaViewControl ViewControl { get; set; }

        public static void GenerateContextMenu(AgendaViewControl viewControl, GridViewMenu contextMenu, SchedulerControl control, Appointment apt) {
            Scheduler = control;
            CurrentAppointment = apt;
            ViewControl = viewControl;
            if(apt != null) {
                contextMenu.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Open appointment", OnOpenCurrentAppointment));
                contextMenu.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Delete appointment", OnDeleteCurrentAppointment, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("AgendViewComponent.Resources.Delete.png"))));                
            }
            
            DevExpress.Utils.Menu.DXMenuItem nextInterval = new DevExpress.Utils.Menu.DXMenuItem("Go to the next month", OnNextInterval, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("AgendViewComponent.Resources.NextArrow.png")));
            nextInterval.BeginGroup = true;
            
            contextMenu.Items.Add(nextInterval);
            contextMenu.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Go to the previous month", OnPreviousInterval, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("AgendViewComponent.Resources.PrevArrow.png"))));
            contextMenu.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Go to the specific date", OnGoToSpecificDate));

            DevExpress.Utils.Menu.DXSubMenuItem switchView = new DevExpress.Utils.Menu.DXSubMenuItem("Change view to");
            switchView.BeginGroup = true;
            contextMenu.Items.Add(switchView);

            if(Scheduler.DayView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Day View", OnSwitchToDayView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("AgendViewComponent.Resources.DayView.png"))));
            if(Scheduler.WorkWeekView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Work Week View", OnSwitchToWorkWeekView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("AgendViewComponent.Resources.WorkWeekView.png"))));
            if(Scheduler.WeekView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Week View", OnSwitchToWeekView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("AgendViewComponent.Resources.WeekView.png"))));
            if(Scheduler.MonthView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Month View", OnSwitchToMonthView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("AgendViewComponent.Resources.MonthView.png"))));
            if(Scheduler.TimelineView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Timeline View", OnSwitchToTimelineView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("AgendViewComponent.Resources.TimelineView.png"))));
            if(Scheduler.GanttView.Enabled) switchView.Items.Add(new DevExpress.Utils.Menu.DXMenuItem("Gantt View", OnSwitchToGanttView, Image.FromStream(Assembly.GetExecutingAssembly().GetManifestResourceStream("AgendViewComponent.Resources.GanttView.png"))));

            if(apt != null) {
                contextMenu.Items.Add(new DevExpress.Utils.Menu.DXMenuItem(ViewControl.ShowResources ? "Show all appointments" : "Group appointments by resources", OnSwitchResourcesVisibility));
            }
        }

        public static void OnOpenCurrentAppointment(object sender, EventArgs e) {
            Scheduler.ShowEditAppointmentForm(CurrentAppointment, CurrentAppointment.IsRecurring);
            ViewControl.InitializeGridControlAppointments();
        }

        public static void OnDeleteCurrentAppointment(object sender, EventArgs e) {
            Scheduler.DeleteAppointment(CurrentAppointment);
            ViewControl.InitializeGridControlAppointments();
        }

        public static void OnGoToSpecificDate(object sender, EventArgs e) {
            GoToDateDialog dateDialog = new GoToDateDialog();
            dateDialog.SelectedDate = AgendaViewDataGenerator.SelectedInterval.Start;
            DialogResult result = dateDialog.ShowDialog();
            if(result == DialogResult.OK) {
                DateTime intervalStart = new DateTime(dateDialog.SelectedDate.Year, dateDialog.SelectedDate.Month, 1);
                AgendaViewDataGenerator.SelectedInterval = new TimeInterval(intervalStart, intervalStart.AddMonths(1));
                ViewControl.InitializeGridControlAppointments();
                ViewControl.GenerateAgndaViewCaption();
            }
        }

        public static void OnNextInterval(object sender, EventArgs e) {
            AgendaViewDataGenerator.SelectedInterval = new TimeInterval(AgendaViewDataGenerator.SelectedInterval.End, AgendaViewDataGenerator.SelectedInterval.End.AddMonths(1));
            ViewControl.InitializeGridControlAppointments();
            ViewControl.GenerateAgndaViewCaption();
        }

        public static void OnPreviousInterval(object sender, EventArgs e) {
            AgendaViewDataGenerator.SelectedInterval = new TimeInterval(AgendaViewDataGenerator.SelectedInterval.Start.AddMonths(-1), AgendaViewDataGenerator.SelectedInterval.Start);
            ViewControl.InitializeGridControlAppointments();
            ViewControl.GenerateAgndaViewCaption();
        }

        public static void OnSwitchToDayView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Day;
            AgendaViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToWorkWeekView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.WorkWeek;
            AgendaViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToWeekView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Week;
            AgendaViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToMonthView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Month;
            AgendaViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToTimelineView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Timeline;
            AgendaViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchToGanttView(object sender, EventArgs e) {
            Scheduler.ActiveViewType = SchedulerViewType.Gantt;
            AgendaViewHelper.SwitchToNormalView();
        }

        public static void OnSwitchResourcesVisibility(object sender, EventArgs e) {
            ViewControl.ShowResources = !ViewControl.ShowResources;
        }
    }
}

File: VB/AgendViewComponent/AgendaViewMenuBuilder.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraScheduler
Imports DevExpress.XtraGrid.Menu
Imports System.Windows.Forms
Imports System.Drawing
Imports System.Reflection

Namespace AgendViewComponent
	Public NotInheritable Class AgendaViewMenuBuilder
        Private Shared privateScheduler As SchedulerControl
		Public Shared Property Scheduler() As SchedulerControl
			Get
				Return privateScheduler
			End Get
			Set(ByVal value As SchedulerControl)
				privateScheduler = value
			End Set
		End Property
        Private Shared privateCurrentAppointment As Appointment
		Public Shared Property CurrentAppointment() As Appointment
			Get
				Return privateCurrentAppointment
			End Get
			Set(ByVal value As Appointment)
				privateCurrentAppointment = value
			End Set
		End Property

        Private Shared privateViewControl As AgendaViewControl
		Public Shared Property ViewControl() As AgendaViewControl
			Get
				Return privateViewControl
			End Get
			Set(ByVal value As AgendaViewControl)
				privateViewControl = value
			End Set
		End Property

		Private Sub New()
		End Sub
        Public Shared Sub GenerateContextMenu(ByVal parViewControl As AgendaViewControl, ByVal contextMenu As GridViewMenu, ByVal control As SchedulerControl, ByVal apt As Appointment)
            Scheduler = control
            CurrentAppointment = apt
            ViewControl = parViewControl
            If apt IsNot Nothing Then
                contextMenu.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Open appointment", AddressOf OnOpenCurrentAppointment))
                contextMenu.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Delete appointment", AddressOf OnDeleteCurrentAppointment, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("Delete.png"))))
            End If

            Dim nextInterval As New DevExpress.Utils.Menu.DXMenuItem("Go to the next month", AddressOf OnNextInterval, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("NextArrow.png")))
            nextInterval.BeginGroup = True

            contextMenu.Items.Add(nextInterval)
            contextMenu.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Go to the previous month", AddressOf OnPreviousInterval, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("PrevArrow.png"))))
            contextMenu.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Go to the specific date", AddressOf OnGoToSpecificDate))

            Dim switchView As New DevExpress.Utils.Menu.DXSubMenuItem("Change view to")
            switchView.BeginGroup = True
            contextMenu.Items.Add(switchView)

            If Scheduler.DayView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Day View", AddressOf OnSwitchToDayView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("DayView.png"))))
            End If
            If Scheduler.WorkWeekView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Work Week View", AddressOf OnSwitchToWorkWeekView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("WorkWeekView.png"))))
            End If
            If Scheduler.WeekView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Week View", AddressOf OnSwitchToWeekView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("WeekView.png"))))
            End If
            If Scheduler.MonthView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Month View", AddressOf OnSwitchToMonthView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("MonthView.png"))))
            End If
            If Scheduler.TimelineView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Timeline View", AddressOf OnSwitchToTimelineView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("TimelineView.png"))))
            End If
            If Scheduler.GanttView.Enabled Then
                switchView.Items.Add(New DevExpress.Utils.Menu.DXMenuItem("Gantt View", AddressOf OnSwitchToGanttView, Image.FromStream(System.Reflection.Assembly.GetExecutingAssembly().GetManifestResourceStream("GanttView.png"))))
            End If

            If apt IsNot Nothing Then
                contextMenu.Items.Add(New DevExpress.Utils.Menu.DXMenuItem(If(viewControl.ShowResources, "Show all appointments", "Group appointments by resources"), AddressOf OnSwitchResourcesVisibility))
            End If
        End Sub

		Public Shared Sub OnOpenCurrentAppointment(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ShowEditAppointmentForm(CurrentAppointment, CurrentAppointment.IsRecurring)
			ViewControl.InitializeGridControlAppointments()
		End Sub

		Public Shared Sub OnDeleteCurrentAppointment(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.DeleteAppointment(CurrentAppointment)
			ViewControl.InitializeGridControlAppointments()
		End Sub

		Public Shared Sub OnGoToSpecificDate(ByVal sender As Object, ByVal e As EventArgs)
			Dim dateDialog As New GoToDateDialog()
			dateDialog.SelectedDate = AgendaViewDataGenerator.SelectedInterval.Start
			Dim result As DialogResult = dateDialog.ShowDialog()
			If result = DialogResult.OK Then
				Dim intervalStart As New DateTime(dateDialog.SelectedDate.Year, dateDialog.SelectedDate.Month, 1)
				AgendaViewDataGenerator.SelectedInterval = New TimeInterval(intervalStart, intervalStart.AddMonths(1))
				ViewControl.InitializeGridControlAppointments()
				ViewControl.GenerateAgndaViewCaption()
			End If
		End Sub

		Public Shared Sub OnNextInterval(ByVal sender As Object, ByVal e As EventArgs)
			AgendaViewDataGenerator.SelectedInterval = New TimeInterval(AgendaViewDataGenerator.SelectedInterval.End, AgendaViewDataGenerator.SelectedInterval.End.AddMonths(1))
			ViewControl.InitializeGridControlAppointments()
			ViewControl.GenerateAgndaViewCaption()
		End Sub

		Public Shared Sub OnPreviousInterval(ByVal sender As Object, ByVal e As EventArgs)
			AgendaViewDataGenerator.SelectedInterval = New TimeInterval(AgendaViewDataGenerator.SelectedInterval.Start.AddMonths(-1), AgendaViewDataGenerator.SelectedInterval.Start)
			ViewControl.InitializeGridControlAppointments()
			ViewControl.GenerateAgndaViewCaption()
		End Sub

		Public Shared Sub OnSwitchToDayView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Day
			AgendaViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToWorkWeekView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.WorkWeek
			AgendaViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToWeekView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Week
			AgendaViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToMonthView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Month
			AgendaViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToTimelineView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Timeline
			AgendaViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchToGanttView(ByVal sender As Object, ByVal e As EventArgs)
			Scheduler.ActiveViewType = SchedulerViewType.Gantt
			AgendaViewHelper.SwitchToNormalView()
		End Sub

		Public Shared Sub OnSwitchResourcesVisibility(ByVal sender As Object, ByVal e As EventArgs)
			ViewControl.ShowResources = Not ViewControl.ShowResources
		End Sub
	End Class
End Namespace
